transformations/synonym_substitution/transformation.py

```diff
@@ -0,0 +1,68 @@
+import random
+import stanza
+
+stanza.download("en")
+import nltk
+
+nltk.download("wordnet")
+from nltk.corpus import wordnet
+from interfaces.SentenceOperation import SentenceOperation
+from tasks.TaskTypes import TaskType
+
+"""
+Base Class for implementing the different input transformations a generation should be robust against.
+"""
+
+
+def synonym_substitution(text, stanza_pipeline, seed=42, prob=0.5):
+    random.seed(seed)
+    upos_wn_dict = {
+        "VERB": "v",
+        "NOUN": "n",
+        "ADV": "r",
+        "ADJ": "s",
+    }
+    result = []
+    doc = stanza_pipeline(text)
+    for sent in doc.sentences:
+        for word_dict in sent.words:
+            word = word_dict.text
+            wn_pos = upos_wn_dict.get(word_dict.upos)
+            if wn_pos is None:
+                result.append(word)
+            else:
+                syns = wordnet.synsets(word, pos=wn_pos)
+                syns = [syn.name().split(".")[0] for syn in syns]
+                syns = [syn for syn in syns if syn != word]
```

Comment:
I can see you've removed the original word from the list of candidate synonyms. But it seems a bit strange that there're still cases where words being transformed to itself, such as in the last test case (American -> American).
```
    {
      "class": "SynonymSubstitution",
      "inputs": {
        "sentence": "A pisco sour is a cocktail typical of South American cuisine. The drink's name is a combination of the word pisco, which is its base liquor, and the term sour, in reference to sour citrus juice and sweetener components."
      },
      "outputs": {
        "sentence": "A pisco sour is a cocktail distinctive of South american cuisine . The drink 's name is a combination of the word pisco , which is its base liquor , and the terminus sourness , in character to sour citrus juice and sweetener components ."
      }
    }
  ]
} 
```
Is it because the synonyms in wordnet are case-sensitive? Or it might be caused by the stanza tokenization (when `random.random()<prob` and `result.append(word)` but the `word` here is lower-cased)?

Reply:
Yeah this is due to the nltk synonyms, which are case-insensitive. For example,

```
>>> syns = wordnet.synsets('American')
>>> syns
[Synset('american.n.01'),
 Synset('american_english.n.01'),
 Synset('american.n.03'),
 Synset('american.a.01'),
 Synset('american.a.02')]
```

Reply:
I see. Would it be good to reject the same word? Because if only one word is selected and replaced with the same word, it does not actually change the sentence.

Reply:
Arguably upper/lower case also changes the semantic meaning of a word. But to disentangle the effect of casing, I have updated the code to make sure that the exact match check is independent of casing.